java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -60,112 +61,165 @@
+        if (conditionState.getBooleanConstraint(conditionState.peek()) == SymbolicBooleanConstraint.FALSE) {
+          result.add(conditionState);
+        } else {
+          conditionState.pop();
+          result.addAll(evaluateCondition(conditionState, tree.rightOperand()));
         }
       }
-      return null;
     }
 
-    @CheckForNull
-    final Symbol.VariableSymbol extractVariableSymbol(ExpressionTree tree) {
-      Tree currentTree = tree;
-      if (isSuperOrThisMemberSelect(tree)) {
-        currentTree = ((MemberSelectExpressionTree) currentTree).identifier();
+    private void evaluateConditionConditionalOr(List<ExecutionState> result, ExecutionState executionState, BinaryExpressionTree tree) {
+      for (ExecutionState conditionState : evaluateCondition(executionState, tree.leftOperand())) {
+        if (conditionState.getBooleanConstraint(conditionState.peek()) == SymbolicBooleanConstraint.TRUE) {
+          result.add(conditionState);
+        } else {
+          conditionState.pop();
+          result.addAll(evaluateCondition(conditionState, tree.rightOperand()));
+        }
       }
-      if (currentTree.is(Tree.Kind.IDENTIFIER)) {
-        IdentifierTree identifierTree = (IdentifierTree) currentTree;
-        Symbol symbol = identifierTree.symbol();
-        if (symbol.isVariableSymbol()) {
-          return (Symbol.VariableSymbol) symbol;
+    }
+
+    private void evaluateConditionIdentifier(List<ExecutionState> result, ExecutionState executionState, Symbol symbol) {
+      SymbolicValue value = executionState.getValue(symbol);
+      switch (executionState.getBooleanConstraint(value)) {
+        case FALSE:
+          result.add(executionState.push(SymbolicValue.BOOLEAN_FALSE));
+          break;
+        case TRUE:
+          result.add(executionState.push(SymbolicValue.BOOLEAN_TRUE));
+          break;
+        default:
+          result.add(new ExecutionState(executionState).setBooleanConstraint(value, SymbolicBooleanConstraint.FALSE).push(SymbolicValue.BOOLEAN_FALSE));
+          result.add(new ExecutionState(executionState).setBooleanConstraint(value, SymbolicBooleanConstraint.TRUE).push(SymbolicValue.BOOLEAN_TRUE));
+      }
+    }
+
+    private void evaluateConditionRelationalOperator(List<ExecutionState> result, ExecutionState executionState, BinaryExpressionTree tree, SymbolicRelation operator) {
+      for (ExecutionState rightState : evaluateWithState(executionState, tree.leftOperand())) {
+        for (ExecutionState operatorState : evaluateWithState(rightState, tree.rightOperand())) {
+          SymbolicValue rightValue = operatorState.pop();
+          SymbolicValue leftValue = operatorState.pop();
+          SymbolicValue operatorResult = executionState.evaluateRelation(leftValue, operator, rightValue);
+          if (operatorResult == SymbolicValue.BOOLEAN_FALSE || operatorResult == SymbolicValue.BOOLEAN_TRUE) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)